File: dart/lib/src/mcp/mcp_contract_runtime.dart
part of 'mcp_server.dart';

const _dartServerName = 'linkedspec-semantic-dart';

final class _McpContractError implements Exception {
  const _McpContractError();
}

late final Map<String, Object?> _mcpBundle = _initializeMcpBundle();

Map<String, Object?> _initializeMcpBundle() {
  if (sha256Hex(utf8.encode(_mcpBundleJson)) != _mcpBundleSha256) {
    throw const _McpContractError();
  }
  final Object? decoded;
  try {
    decoded = jsonDecode(_mcpBundleJson);
  } on Object {
    throw const _McpContractError();
  }
  final root = _jsonMap(decoded);
  if (root == null || root['binding_format'] != _mcpBindingFormat) {
    throw const _McpContractError();
  }
  final semanticPayloads = _jsonMap(root['semantic_payloads']);
  final payloads = semanticPayloads?['payloads'];
  if (payloads is! List) {
    throw const _McpContractError();
  }
  final ids = <String>{};
  for (final payload in payloads) {
    final row = _jsonMap(payload);
    final id = row?['id'];
    if (id is! String || !ids.add(id)) {
      throw const _McpContractError();
    }
  }
  return root;
}

Map<String, Object?> _mcpContract() {
  final contract = _mcpBundle['contract'];
  final result = _jsonMap(_copyJson(contract));
  if (result == null) {
    throw const _McpContractError();
  }
  return result;
}

String _mcpProtocolVersion() {
  final contract = _jsonMap(_mcpBundle['contract']);
  final version = contract?['protocol_version'];
  if (version is! String) {
    throw const _McpContractError();
  }
  return version;
}

Map<String, Object?>? _mcpFrame(String id) {
  final frames = _jsonMap(_mcpBundle['canonical_frames']);
  if (frames == null || !frames.containsKey(id)) {
    return null;
  }
  return _jsonMap(_copyJson(frames[id]));
}

String _mcpCanonicalJson(Object? value) {
  try {
    return jsonEncode(_canonicalJsonValue(_copyJson(value)));
  } on Object {
    throw const _McpContractError();
  }
}

Object? _canonicalJsonValue(Object? value) {
  if (value is List) {
    return <Object?>[for (final child in value) _canonicalJsonValue(child)];
  }
  if (value is Map) {
    final sorted = SplayTreeMap<String, Object?>();
    for (final entry in value.entries) {
      if (entry.key is! String) {
        throw const _McpContractError();
      }
      sorted[entry.key as String] = _canonicalJsonValue(entry.value);
    }
    return sorted;
  }
  if (value is double && !value.isFinite) {
    throw const _McpContractError();
  }
  if (value == null ||
      value is bool ||
      value is String ||
      value is int ||
      value is double) {
    return value;
  }
  throw const _McpContractError();
}

bool _mcpValidateNamed(String name, Object? value) {
  final schema = _jsonMap(_mcpBundle['schema']);
  final definitions = _jsonMap(schema?['\$defs']);
  final definition = definitions?[name];
  return definition != null && _mcpMatches(value, definition, schema!, 0);
}

bool _mcpValidateFrame(Object? value) {
  final schema = _jsonMap(_mcpBundle['schema']);
  return schema != null && _mcpMatches(value, schema, schema, 0);
}

bool _mcpMatches(
  Object? value,
  Object? schema,
  Map<String, Object?> root,
  int depth,
) {
  try {
    _mcpValidate(value, schema, root, depth);
    return true;
  } on Object {
    return false;
  }
}

void _mcpValidate(
  Object? value,
  Object? schemaValue,
  Map<String, Object?> root,
  int depth,
) {
  if (depth > 256) {
    throw const _McpContractError();
  }
  if (schemaValue is bool) {
    if (!schemaValue) {
      throw const _McpContractError();
    }
    return;
  }
  final schema = _jsonMap(schemaValue);
  if (schema == null) {
    throw const _McpContractError();
  }

  final reference = schema['\$ref'];
  if (reference != null) {
    if (reference is! String || !reference.startsWith('#/\$defs/')) {
      throw const _McpContractError();
    }
    final name = reference.substring('#/\$defs/'.length);
    if (name.isEmpty || name.contains('/')) {
      throw const _McpContractError();
    }
    final definitions = _jsonMap(root['\$defs']);
    final definition = definitions?[name];
    if (definition == null) {
      throw const _McpContractError();
    }
    _mcpValidate(value, definition, root, depth + 1);
  }

  final type = schema['type'];
  if (type != null) {
    final matches = switch (type) {
      String kind => _mcpHasType(value, kind),
      List<Object?> kinds => kinds.whereType<String>().any(
        (kind) => _mcpHasType(value, kind),
      ),
      _ => false,
    };
    if (!matches) {
      throw const _McpContractError();
    }
  }
  if (schema.containsKey('const') && !_jsonEqual(value, schema['const'])) {
    throw const _McpContractError();
  }
  final allowed = schema['enum'];
  if (allowed != null &&
      (allowed is! List || !allowed.any((item) => _jsonEqual(value, item)))) {
    throw const _McpContractError();
  }
  final oneOf = schema['oneOf'];
  if (oneOf != null) {
    if (oneOf is! List ||
        oneOf
                .where((branch) => _mcpMatches(value, branch, root, depth + 1))
                .length !=
            1) {
      throw const _McpContractError();
    }
  }
  final allOf = schema['allOf'];
  if (allOf != null) {
    if (allOf is! List) {
      throw const _McpContractError();
    }
    for (final branch in allOf) {
      _mcpValidate(value, branch, root, depth + 1);
    }
  }

  final object = _jsonMap(value);
  if (object != null) {
    _mcpValidateObject(object, schema, root, depth);
  }
  if (value is List) {
    _mcpValidateArray(value, schema, root, depth);
  }
  if (value is String) {
    _mcpValidateString(value, schema);
  }
  if (value is int) {
    _mcpValidateInteger(value, schema);
  }
}

void _mcpValidateObject(
  Map<String, Object?> value,
  Map<String, Object?> schema,
  Map<String, Object?> root,
  int depth,
) {
  final maximum = schema['maxProperties'];
  if (maximum is int && value.length > maximum) {
    throw const _McpContractError();
  }
  final required = schema['required'];
  if (required != null) {
    if (required is! List ||
        required.whereType<String>().any((name) => !value.containsKey(name))) {
      throw const _McpContractError();
    }
  }
  final properties = _jsonMap(schema['properties']);
  for (final entry in value.entries) {
    final propertyNames = schema['propertyNames'];
    if (propertyNames != null) {
      _mcpValidate(entry.key, propertyNames, root, depth + 1);
    }
    if (properties?.containsKey(entry.key) == true) {
      _mcpValidate(entry.value, properties![entry.key], root, depth + 1);
      continue;
    }
    final additional = schema['additionalProperties'];
    if (additional == false) {
      throw const _McpContractError();
    }
    if (additional is Map || additional is bool) {
      _mcpValidate(entry.value, additional, root, depth + 1);
    }
  }
}

void _mcpValidateArray(
  List<Object?> value,
  Map<String, Object?> schema,
  Map<String, Object?> root,
  int depth,
) {
  final prefix = schema['prefixItems'];
  final prefixItems = prefix is List ? prefix : const <Object?>[];
  final prefixCount = min(value.length, prefixItems.length);
  for (var index = 0; index < prefixCount; index += 1) {
    _mcpValidate(value[index], prefixItems[index], root, depth + 1);
  }
  final items = schema['items'];
  if (items != null) {
    for (var index = prefixItems.length; index < value.length; index += 1) {
      _mcpValidate(value[index], items, root, depth + 1);
    }
  }
  final minimum = schema['minItems'];
  final maximum = schema['maxItems'];
  if ((minimum is int && value.length < minimum) ||
      (maximum is int && value.length > maximum)) {
    throw const _McpContractError();
  }
}

void _mcpValidateString(String value, Map<String, Object?> schema) {
  final length = value.runes.length;
  final minimum = schema['minLength'];
  final maximum = schema['maxLength'];
  final maximumBytes = schema['x-linkedspec-maxUtf8Bytes'];
  if ((minimum is int && length < minimum) ||
      (maximum is int && length > maximum) ||
      (maximumBytes is int && utf8.encode(value).length > maximumBytes)) {
    throw const _McpContractError();
  }
  final pattern = schema['pattern'];
  if (pattern != null) {
    final matches = switch (pattern) {
      r'^[A-Za-z0-9_-]{43}$' =>
        value.length == 43 && RegExp(r'^[A-Za-z0-9_-]{43}$').hasMatch(value),
      r'^sha256:[0-9a-f]{64}$' =>
        value.length == 71 && RegExp(r'^sha256:[0-9a-f]{64}$').hasMatch(value),
      _ => false,
    };
    if (!matches) {
      throw const _McpContractError();
    }
  }
  if (schema['format'] == 'uri' &&
      !RegExp(r'^[A-Za-z][A-Za-z0-9+.-]*:').hasMatch(value)) {
    throw const _McpContractError();
  }
}

void _mcpValidateInteger(int value, Map<String, Object?> schema) {
  final minimum = schema['minimum'];
  final maximum = schema['maximum'];
  if ((minimum is int && value < minimum) ||
      (maximum is int && value > maximum)) {
    throw const _McpContractError();
  }
}

bool _mcpHasType(Object? value, String kind) => switch (kind) {
  'null' => value == null,
  'boolean' => value is bool,
  'array' => value is List,
  'object' => _jsonMap(value) != null,
  'string' => value is String,
  'integer' => value is int,
  'number' => value is num && (value is! double || value.isFinite),
  _ => false,
};

bool _jsonEqual(Object? left, Object? right) {
  if (left is num && right is num) {
    return left.runtimeType == right.runtimeType && left == right;
  }
  if (left is List && right is List) {
    return left.length == right.length &&
        Iterable<int>.generate(
          left.length,
        ).every((index) => _jsonEqual(left[index], right[index]));
  }
  final leftMap = _jsonMap(left);
  final rightMap = _jsonMap(right);
  if (leftMap != null || rightMap != null) {
    return leftMap != null &&
        rightMap != null &&
        leftMap.length == rightMap.length &&
        leftMap.entries.every(
          (entry) =>
              rightMap.containsKey(entry.key) &&
              _jsonEqual(entry.value, rightMap[entry.key]),
        );
  }
  return left == right;
}

Map<String, Object?> _mcpDiscoverResponse(Object id) {
  final response = _mcpFrame('discover_response_dart');
  if (response == null) {
    throw const _McpContractError();
  }
  response['id'] = _copyJson(id);
  return response;
}

Map<String, Object?> _mcpToolsListResponse(Object id) {
  final response = _mcpFrame('tools_list_response_perl');
  if (response == null) {
    throw const _McpContractError();
  }
  _setDartServerName(response);
  response['id'] = _copyJson(id);
  return response;
}

Map<String, Object?> _mcpToolSuccessResponse(
  Object id,
  Map<String, Object?> payload,
) {
  if (!_mcpValidateNamed('semanticQueryResponse', payload)) {
    throw const _McpContractError();
  }
  final response = _mcpFrame('capabilities_call_response');
  if (response == null) {
    throw const _McpContractError();
  }
  _setDartServerName(response);
  response['id'] = _copyJson(id);
  final result = _requiredMap(response['result']);
  result['structuredContent'] = _copyJson(payload);
  final content = result['content'];
  if (content is! List || content.isEmpty) {
    throw const _McpContractError();
  }
  final first = _requiredMap(content.first);
  first['text'] = _mcpCanonicalJson(payload);
  return response;
}

Map<String, Object?> _mcpToolErrorResponse(
  Object id, {
  required bool unavailable,
}) {
  final response = _mcpFrame('policy_denied_response');
  if (response == null) {
    throw const _McpContractError();
  }
  _setDartServerName(response);
  if (unavailable) {
    final unavailableResponse = _mcpFrame('handle_unavailable_response');
    final unavailableResult = _jsonMap(unavailableResponse?['result']);
    final result = _requiredMap(response['result']);
    if (unavailableResult == null) {
      throw const _McpContractError();
    }
    result['content'] = _copyJson(unavailableResult['content']);
  }
  response['id'] = _copyJson(id);
  return response;
}

Map<String, Object?> _mcpJsonRpcError(
  Object? id,
  String kind, {
  String? requested,
}) {
  final contract = _jsonMap(_mcpBundle['contract']);
  if (contract == null) {
    throw const _McpContractError();
  }
  int? code;
  String? message;
  if (kind == 'legacy_initialize') {
    final legacy = _jsonMap(contract['legacy_diagnostic']);
    code = legacy?['code'] as int?;
    message = legacy?['message'] as String?;
  } else {
    final owner = switch (kind) {
      'parse_error' => 'invalid_utf8',
      'invalid_request' => 'invalid_envelope',
      'method_not_found' => 'unknown_method',
      'invalid_params' => 'method_params',
      'internal_error' => 'sanitized_unexpected_failure',
      'unsupported_version' => 'unsupported_protocol_version',
      _ => null,
    };
    final rows = contract['json_rpc_errors'];
    if (owner == null || rows is! List) {
      throw const _McpContractError();
    }
    for (final candidate in rows) {
      final row = _jsonMap(candidate);
      final owns = row?['owns'];
      if (owns is List && owns.contains(owner)) {
        code = row?['code'] as int?;
        message = row?['message'] as String?;
        break;
      }
    }
  }
  if (code == null || message == null) {
    throw const _McpContractError();
  }
  final error = <String, Object?>{'code': code, 'message': message};
  if (kind == 'unsupported_version') {
    error['data'] = {
      'requested': requested ?? '',
      'supported': [_mcpProtocolVersion()],
    };
  }
  return {'jsonrpc': '2.0', 'id': _copyJson(id), 'error': error};
}

void _setDartServerName(Map<String, Object?> response) {
  final result = _requiredMap(response['result']);
  final metadata = _requiredMap(result['_meta']);
  final server = _requiredMap(metadata['io.modelcontextprotocol/serverInfo']);
  server['name'] = _dartServerName;
}

File: julia/src/mcp/McpContractRuntime.jl
# Frozen runtime over the generated, filesystem-free MCP contract binding.

struct _McpContractError <: Exception end

const _MCP_REF_PREFIX = "#/\$defs/"
const _MCP_HANDLE_PATTERN = r"^[A-Za-z0-9_-]{43}$"
const _MCP_DIGEST_PATTERN = r"^sha256:[0-9a-f]{64}$"
const _MCP_URI_PATTERN = r"^[A-Za-z][A-Za-z0-9+.-]*:"

function _mcp_initialize_bundle()
    raw = try
        Base64.base64decode(_MCP_BUNDLE_BASE64)
    catch
        throw(_McpContractError())
    end
    bytes2hex(SHA.sha256(raw)) == _MCP_BUNDLE_SHA256 || throw(_McpContractError())
    text = String(raw)
    isvalid(text) || throw(_McpContractError())
    root = try
        JSON3.read(text, Dict{String,Any})
    catch
        throw(_McpContractError())
    end
    get(root, "binding_format", nothing) == _MCP_BINDING_FORMAT ||
        throw(_McpContractError())
    semantic = _mcp_as_object(get(root, "semantic_payloads", nothing))
    payloads = semantic === nothing ? nothing : get(semantic, "payloads", nothing)
    payloads isa AbstractVector || throw(_McpContractError())
    ids = Set{String}()
    for payload in payloads
        row = _mcp_as_object(payload)
        id = row === nothing ? nothing : get(row, "id", nothing)
        id isa AbstractString || throw(_McpContractError())
        normalized = String(id)
        normalized in ids && throw(_McpContractError())
        push!(ids, normalized)
    end
    return root
end

function _mcp_contract()
    contract = _mcp_copy_json(get(_MCP_BUNDLE, "contract", nothing))
    contract isa Dict{String,Any} || throw(_McpContractError())
    return contract
end

function _mcp_protocol_version()
    contract = _mcp_as_object(get(_MCP_BUNDLE, "contract", nothing))
    version = contract === nothing ? nothing : get(contract, "protocol_version", nothing)
    version isa AbstractString || throw(_McpContractError())
    return String(version)
end

function _mcp_frame(name::AbstractString)
    frames = _mcp_as_object(get(_MCP_BUNDLE, "canonical_frames", nothing))
    frames === nothing && throw(_McpContractError())
    haskey(frames, name) || return nothing
    value = _mcp_copy_json(frames[name])
    value isa Dict{String,Any} || throw(_McpContractError())
    return value
end

function _mcp_canonical_json(value)
    copied = _mcp_copy_json(value)
    io = IOBuffer()
    _mcp_write_canonical(io, copied)
    return String(take!(io))
end

function _mcp_write_canonical(io::IO, value)
    if value isa AbstractDict
        keys_sorted = sort!(collect(keys(value)))
        write(io, UInt8('{'))
        for (index, key) in enumerate(keys_sorted)
            index == 1 || write(io, UInt8(','))
            write(io, JSON3.write(String(key)))
            write(io, UInt8(':'))
            _mcp_write_canonical(io, value[key])
        end
        write(io, UInt8('}'))
    elseif value isa AbstractVector
        write(io, UInt8('['))
        for (index, child) in enumerate(value)
            index == 1 || write(io, UInt8(','))
            _mcp_write_canonical(io, child)
        end
        write(io, UInt8(']'))
    else
        write(io, JSON3.write(value))
    end
    return nothing
end

function _mcp_validate_named(name::AbstractString, value)
    schema = _mcp_as_object(get(_MCP_BUNDLE, "schema", nothing))
    definitions = schema === nothing ? nothing : _mcp_as_object(get(schema, "\$defs", nothing))
    definition = definitions === nothing ? nothing : get(definitions, name, nothing)
    definition === nothing && return false
    return _mcp_matches(value, definition, schema, 0)
end

function _mcp_validate_frame(value)
    schema = _mcp_as_object(get(_MCP_BUNDLE, "schema", nothing))
    schema === nothing && return false
    return _mcp_matches(value, schema, schema, 0)
end

function _mcp_matches(value, schema, root::AbstractDict, depth::Int)
    try
        _mcp_validate(value, schema, root, depth)
        return true
    catch
        return false
    end
end

function _mcp_validate(value, schema_value, root::AbstractDict, depth::Int)
    depth <= 256 || throw(_McpContractError())
    if schema_value isa Bool
        schema_value || throw(_McpContractError())
        return nothing
    end
    schema = _mcp_as_object(schema_value)
    schema === nothing && throw(_McpContractError())

    reference = get(schema, "\$ref", nothing)
    if reference !== nothing
        reference isa AbstractString || throw(_McpContractError())
        startswith(reference, _MCP_REF_PREFIX) || throw(_McpContractError())
        reference_text = String(reference)
        length(reference_text) > length(_MCP_REF_PREFIX) || throw(_McpContractError())
        name = reference_text[(length(_MCP_REF_PREFIX) + 1):end]
        occursin('/', name) && throw(_McpContractError())
        definitions = _mcp_as_object(get(root, "\$defs", nothing))
        definition = definitions === nothing ? nothing : get(definitions, name, nothing)
        definition === nothing && throw(_McpContractError())
        _mcp_validate(value, definition, root, depth + 1)
    end

    type_value = get(schema, "type", nothing)
    if type_value !== nothing
        matches = if type_value isa AbstractString
            _mcp_has_type(value, type_value)
        elseif type_value isa AbstractVector
            any(kind -> kind isa AbstractString && _mcp_has_type(value, kind), type_value)
        else
            false
        end
        matches || throw(_McpContractError())
    end
    if haskey(schema, "const") && !_mcp_json_equal(value, schema["const"])
        throw(_McpContractError())
    end
    allowed = get(schema, "enum", nothing)
    if allowed !== nothing
        allowed isa AbstractVector || throw(_McpContractError())
        any(expected -> _mcp_json_equal(value, expected), allowed) ||
            throw(_McpContractError())
    end
    one_of = get(schema, "oneOf", nothing)
    if one_of !== nothing
        one_of isa AbstractVector || throw(_McpContractError())
        count(branch -> _mcp_matches(value, branch, root, depth + 1), one_of) == 1 ||
            throw(_McpContractError())
    end
    all_of = get(schema, "allOf", nothing)
    if all_of !== nothing
        all_of isa AbstractVector || throw(_McpContractError())
        for branch in all_of
            _mcp_validate(value, branch, root, depth + 1)
        end
    end

    object = _mcp_as_object(value)
    object === nothing || _mcp_validate_object(object, schema, root, depth)
    value isa AbstractVector && _mcp_validate_array(value, schema, root, depth)
    value isa AbstractString && _mcp_validate_string(value, schema)
    _mcp_is_integer(value) && _mcp_validate_integer(value, schema)
    return nothing
end

function _mcp_validate_object(value::AbstractDict, schema::AbstractDict, root, depth)
    maximum = get(schema, "maxProperties", nothing)
    maximum isa Integer && length(value) > maximum && throw(_McpContractError())
    required = get(schema, "required", nothing)
    if required !== nothing
        required isa AbstractVector || throw(_McpContractError())
        all(name -> name isa AbstractString && haskey(value, name), required) ||
            throw(_McpContractError())
    end
    properties = _mcp_as_object(get(schema, "properties", nothing))
    for (name, child) in value
        property_names = get(schema, "propertyNames", nothing)
        property_names === nothing || _mcp_validate(name, property_names, root, depth + 1)
        if properties !== nothing && haskey(properties, name)
            _mcp_validate(child, properties[name], root, depth + 1)
            continue
        end
        additional = get(schema, "additionalProperties", nothing)
        additional === false && throw(_McpContractError())
        if additional isa AbstractDict || additional isa Bool
            _mcp_validate(child, additional, root, depth + 1)
        end
    end
    return nothing
end

function _mcp_validate_array(value::AbstractVector, schema::AbstractDict, root, depth)
    prefix_value = get(schema, "prefixItems", nothing)
    prefix = prefix_value isa AbstractVector ? prefix_value : ()
    for index in 1:min(length(value), length(prefix))
        _mcp_validate(value[index], prefix[index], root, depth + 1)
    end
    items = get(schema, "items", nothing)
    if items !== nothing
        for index in (length(prefix) + 1):length(value)
            _mcp_validate(value[index], items, root, depth + 1)
        end
    end
    minimum = get(schema, "minItems", nothing)
    maximum = get(schema, "maxItems", nothing)
    minimum isa Integer && length(value) < minimum && throw(_McpContractError())
    maximum isa Integer && length(value) > maximum && throw(_McpContractError())
    return nothing
end

function _mcp_validate_string(value::AbstractString, schema::AbstractDict)
    isvalid(value) || throw(_McpContractError())
    minimum = get(schema, "minLength", nothing)
    maximum = get(schema, "maxLength", nothing)
    maximum_bytes = get(schema, "x-linkedspec-maxUtf8Bytes", nothing)
    minimum isa Integer && length(value) < minimum && throw(_McpContractError())
    maximum isa Integer && length(value) > maximum && throw(_McpContractError())
    maximum_bytes isa Integer && ncodeunits(value) > maximum_bytes && throw(_McpContractError())
    pattern = get(schema, "pattern", nothing)
    if pattern !== nothing
        matches = if pattern == "^[A-Za-z0-9_-]{43}\$"
            occursin(_MCP_HANDLE_PATTERN, value)
        elseif pattern == "^sha256:[0-9a-f]{64}\$"
            occursin(_MCP_DIGEST_PATTERN, value)
        else
            false
        end
        matches || throw(_McpContractError())
    end
    get(schema, "format", nothing) == "uri" &&
        !occursin(_MCP_URI_PATTERN, value) && throw(_McpContractError())
    return nothing
end

function _mcp_validate_integer(value::Integer, schema::AbstractDict)
    minimum = get(schema, "minimum", nothing)
    maximum = get(schema, "maximum", nothing)
    minimum isa Integer && value < minimum && throw(_McpContractError())
    maximum isa Integer && value > maximum && throw(_McpContractError())
    return nothing
end

function _mcp_has_type(value, kind::AbstractString)
    kind == "null" && return value === nothing
    kind == "boolean" && return value isa Bool
    kind == "array" && return value isa AbstractVector
    kind == "object" && return _mcp_as_object(value) !== nothing
    kind == "string" && return value isa AbstractString && isvalid(value)
    kind == "integer" && return _mcp_is_integer(value)
    kind == "number" && return _mcp_is_number(value)
    return false
end

_mcp_is_integer(value) = value isa Integer && !(value isa Bool)
_mcp_is_number(value) =
    value isa Real && !(value isa Bool) && (!(value isa AbstractFloat) || isfinite(value))

function _mcp_json_equal(left, right)
    if _mcp_is_number(left) || _mcp_is_number(right)
        return _mcp_is_number(left) && _mcp_is_number(right) &&
               typeof(left) == typeof(right) && left == right
    end
    if left isa AbstractVector || right isa AbstractVector
        return left isa AbstractVector && right isa AbstractVector &&
               length(left) == length(right) &&
               all(index -> _mcp_json_equal(left[index], right[index]), eachindex(left))
    end
    left_object = _mcp_as_object(left)
    right_object = _mcp_as_object(right)
    if left_object !== nothing || right_object !== nothing
        return left_object !== nothing && right_object !== nothing &&
               length(left_object) == length(right_object) &&
               all(
            pair -> haskey(right_object, first(pair)) &&
                    _mcp_json_equal(last(pair), right_object[first(pair)]),
            left_object,
        )
    end
    return left == right
end

function _mcp_discover_response(id)
    response = _mcp_frame("discover_response_julia")
    response === nothing && throw(_McpContractError())
    response["id"] = _mcp_copy_json(id)
    return response
end

function _mcp_tools_list_response(id)
    response = _mcp_frame("tools_list_response_perl")
    response === nothing && throw(_McpContractError())
    _mcp_set_server_name!(response)
    response["id"] = _mcp_copy_json(id)
    return response
end

function _mcp_tool_success_response(id, payload::Dict{String,Any})
    _mcp_validate_named("semanticQueryResponse", payload) || throw(_McpContractError())
    response = _mcp_frame("capabilities_call_response")
    response === nothing && throw(_McpContractError())
    _mcp_set_server_name!(response)
    response["id"] = _mcp_copy_json(id)
    result = _mcp_required_object(response["result"])
    result["structuredContent"] = _mcp_copy_json(payload)
    content = get(result, "content", nothing)
    content isa AbstractVector && !isempty(content) || throw(_McpContractError())
    first = _mcp_required_object(content[1])
    first["text"] = _mcp_canonical_json(payload)
    return response
end

function _mcp_tool_error_response(id; unavailable::Bool)
    response = _mcp_frame("policy_denied_response")
    response === nothing && throw(_McpContractError())
    _mcp_set_server_name!(response)
    if unavailable
        unavailable_response = _mcp_frame("handle_unavailable_response")
        unavailable_result = unavailable_response === nothing ? nothing :
                             _mcp_as_object(get(unavailable_response, "result", nothing))
        unavailable_result === nothing && throw(_McpContractError())
        result = _mcp_required_object(response["result"])
        result["content"] = _mcp_copy_json(unavailable_result["content"])
    end
    response["id"] = _mcp_copy_json(id)
    return response
end

function _mcp_json_rpc_error(id, kind::AbstractString; requested = nothing)
    contract = _mcp_as_object(get(_MCP_BUNDLE, "contract", nothing))
    contract === nothing && throw(_McpContractError())
    code = nothing
    message = nothing
    if kind == "legacy_initialize"
        legacy = _mcp_as_object(get(contract, "legacy_diagnostic", nothing))
        code = legacy === nothing ? nothing : get(legacy, "code", nothing)
        message = legacy === nothing ? nothing : get(legacy, "message", nothing)
    else
        owner = get(
            Dict(
                "parse_error" => "invalid_utf8",
                "invalid_request" => "invalid_envelope",
                "method_not_found" => "unknown_method",
                "invalid_params" => "method_params",
                "internal_error" => "sanitized_unexpected_failure",
                "unsupported_version" => "unsupported_protocol_version",
            ),
            String(kind),
            nothing,
        )
        rows = get(contract, "json_rpc_errors", nothing)
        owner === nothing || rows isa AbstractVector || throw(_McpContractError())
        for candidate in rows
            row = _mcp_as_object(candidate)
            owns = row === nothing ? nothing : get(row, "owns", nothing)
            if owns isa AbstractVector && owner in owns
                code = get(row, "code", nothing)
                message = get(row, "message", nothing)
                break
            end
        end
    end
    code isa Integer && message isa AbstractString || throw(_McpContractError())
    error = Dict{String,Any}("code" => Int(code), "message" => String(message))
    if kind == "unsupported_version"
        error["data"] = Dict{String,Any}(
            "requested" => requested === nothing ? "" : String(requested),
            "supported" => Any[_mcp_protocol_version()],
        )
    end
    return Dict{String,Any}(
        "jsonrpc" => "2.0",
        "id" => _mcp_copy_json(id),
        "error" => error,
    )
end

function _mcp_set_server_name!(response::Dict{String,Any})
    result = _mcp_required_object(response["result"])
    metadata = _mcp_required_object(result["_meta"])
    server = _mcp_required_object(metadata["io.modelcontextprotocol/serverInfo"])
    server["name"] = _MCP_SERVER_NAME
    return response
end

function _mcp_as_object(value)
    value isa Dict{String,Any} && return value
    value isa AbstractDict || return nothing
    result = Dict{String,Any}()
    for (key, child) in value
        key isa AbstractString || return nothing
        normalized = String(key)
        haskey(result, normalized) && return nothing
        result[normalized] = child
    end
    return result
end

function _mcp_required_object(value)
    result = _mcp_as_object(value)
    result === nothing && throw(_McpContractError())
    return result
end

function _mcp_copy_json(value, depth::Int = 0, active = IdDict{Any,Nothing}())
    depth <= 256 || throw(_McpContractError())
    if value === nothing || value isa Bool
        return value
    elseif value isa AbstractString
        isvalid(value) || throw(_McpContractError())
        return String(value)
    elseif _mcp_is_integer(value)
        return value
    elseif value isa AbstractFloat
        isfinite(value) || throw(_McpContractError())
        return value
    elseif value isa Tuple
        return Any[_mcp_copy_json(child, depth + 1, active) for child in value]
    elseif value isa AbstractVector
        haskey(active, value) && throw(_McpContractError())
        active[value] = nothing
        try
            return Any[_mcp_copy_json(child, depth + 1, active) for child in value]
        finally
            delete!(active, value)
        end
    elseif value isa AbstractDict
        haskey(active, value) && throw(_McpContractError())
        active[value] = nothing
        try
            result = Dict{String,Any}()
            for (key, child) in value
                key isa Union{AbstractString,Symbol} || throw(_McpContractError())
                name = String(key)
                isvalid(name) || throw(_McpContractError())
                haskey(result, name) && throw(_McpContractError())
                result[name] = _mcp_copy_json(child, depth + 1, active)
            end
            return result
        finally
            delete!(active, value)
        end
    end
    throw(_McpContractError())
end

const _MCP_BUNDLE = _mcp_initialize_bundle()
